File: lib/data_structure/Stack.rb
# bounded Stack
class Stack

  def initialize size
    @stack = Array.new size
    @top = -1
    @upper_bound = size-1
    @lower_bound = 0
  end

  def push num
    raise "out of bounds" if @top+1 > @upper_bound
    @stack[@top+=1] = num
  end

  def pop
    raise "can't pop empty stack" if @top < @lower_bound
    @top -= 1
  end

  def peek
    @stack[@top]
  end
end

File: public/javascripts/dataStructure/Stack.js
// bounded Stack
function Stack(size) {
  this.upperBound = size-1;
  this.lowerBound = -1;
  this.stack = new Array(size);
  this.top = -1;
}

Stack.prototype.push = function(value) {
  if(this.top < this.upperBound) {
    this.top++;
    this.stack[this.top] = value;
  } else {
    throw new Error('Stack is full');
  }
};

Stack.prototype.pop = function() {
  if(this.top > this.lowerBound) {
    this.stack[top] = null;
    this.top--;
  } else {
    throw new Error('Stack is empty');
  }
};

Stack.prototype.peek = function() {
  return this.stack[this.top];
};
